cranelift/codegen/src/data_value.rs

```diff
@@ -101,12 +101,12 @@ impl DataValue {
         }
     }
 
-    /// Write a [DataValue] to a slice.
+    /// Write a [DataValue] to a slice in native-endian byte order.
     ///
     /// # Panics:
     ///
     /// Panics if the slice does not have enough space to accommodate the [DataValue]
-    pub fn write_to_slice(&self, dst: &mut [u8]) {
+    pub fn write_to_slice_ne(&self, dst: &mut [u8]) {
```

Comment:
@afonso360 you posited theses functions more as a question, so let me know if you don't think it will suffice.

Also, this could probably be DRY'd up with a macro, but my macrofu is not there yet. Let me know if I should go and do it though.

Reply:
Yeah, mostly because they match what we have in the rust standard library so it seemed natural to keep that naming convention.

This does involve a lot of repetitive code, and I also don't see a way out of this without creating a macro. @jameysharp any ideas here?

Reply:
Huh, yeah, this is tricky! As much as I dislike the duplication, I suspect a macro won't make the code more clear.

One question first: why aren't vectors byte-swapped in this implementation? Is there something I'm missing that makes them special?

Here's an alternative way to split this up. I'm not claiming it's the right choice, but it's an alternative we can discuss.

First, write a function that takes a `DataValue` and returns a `DataValue`, but with its contents unconditionally byte-swapped. In other words, if you have a `DataValue::I32(i)`, return `DataValue::I32(i.swap_bytes())`. For vectors, maybe we should be calling `reverse` on the arrays?

Next, write a pair of functions on `DataValue`, where one calls the above byte-swap function when run on a big-endian system, and the other calls it when run on a little-endian system. See the example code in [`u32::to_be`](https://doc.rust-lang.org/std/primitive.u32.html#method.to_be) and `to_le` for the pattern to follow. Note that `to_be` and `from_be` are functionally identical (both functions only byte-swap on little-endian systems), and similarly for `to_le` and `from_le`. So we don't actually need `to` and `from` variants for each endianness; just one will do.

Finally, keep `write_to_slice_ne` and `read_from_slice_ne` as they are here. The `_le` variants should call the `_ne` variants, but pass the `DataValue` through the method that only byte-swaps on big-endian systems. Similarly, the `_be` variants call the method that byte-swaps on little-endian systems. When reading, the swap happens after the native-endian read; when writing, the swap happens before the native-endian write.

That reduces this from six copies of this match on all of `DataValue`'s variants, to three. That's still two more than I'd prefer, but I think it's pretty reasonable.

Reply:
Thank you for the input!

No, vectors are not special. It seems I just slipped up. Thanks for pointing it out.

As for your proposal, I would be inclined to agree with you. I am in favour of cutting it down so as to make it more manageable and catch errors like the one above.

Reply:
I used the Rust core lib as inspiration, and modelled it quite closely. However, I am unsure about whether to change the `write_to_slice_*` and `read_from_slice_*` methods to move instead of borrow. The result is having to copy data in the write calls. Let me know if this is not desirable.

Reply:
I didn't think about needing to clone! But I think the way you've done it is fine. For one thing, this is only 32 bytes if I counted correctly, so it's not that much to copy. LLVM might even optimize away the copies. But more importantly, this is primarily used in the interpreter, which is already not expected to be super fast.

Reply:
Great, thanks again!

I figured it may not be that big a deal, but I drew attention to it just in case. I will keep this in mind going forward though.